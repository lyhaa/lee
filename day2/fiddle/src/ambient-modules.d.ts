declare module 'fix-path';
declare module 'namor';
