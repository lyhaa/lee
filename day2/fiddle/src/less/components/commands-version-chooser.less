button.version-chooser.bp3-button {
  border-radius: 3px 0 0 3px;
}
