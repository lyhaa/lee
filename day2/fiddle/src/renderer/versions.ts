import { ElectronVersion, ElectronVersionSource, ElectronVersionState, NpmVersion } from '../interfaces';
import { normalizeVersion } from '../utils/normalize-version';

export const enum ElectronReleaseChannel {
  stable = 'Stable',
  beta = 'Beta',
  nightly = 'Nightly',
  unsupported = 'Unsupported'
}

/**
 * Returns a sensible default version string.
 *
 * @param {Array<ElectronVersion>} knownVersions
 * @returns {string}
 */
export function getDefaultVersion(
  knownVersions: Array<ElectronVersion> = []
): string {
  const ls = localStorage.getItem('version');

  if (ls && knownVersions && knownVersions.find(({ version }) => version === ls)) {
    return ls;
  }

  // Self-heal: Version not formated correctly
  const normalized = ls && normalizeVersion(ls);
  if (normalized) {
    if (knownVersions && knownVersions.find(({ version }) => version === normalized)) {
      return normalized;
    }
  }

  // Alright, the first version?
  const last = knownVersions && knownVersions[knownVersions.length - 1];
  if (last) {
    return last.version;
  }

  // Report error
  throw new Error('Corrupted version data');
}

/**
 * Return the release channel for a given input
 * version.
 *
 * @param {NpmVersion | string} input
 * @returns {ElectronReleaseChannel}
 */
export function getReleaseChannel(
  input: NpmVersion | string
): ElectronReleaseChannel {

  const tag = (typeof input === 'string') ? input : (input.version || '');

  if (tag.includes('beta')) {
    return ElectronReleaseChannel.beta;
  }

  if (tag.includes('nightly')) {
    return ElectronReleaseChannel.nightly;
  }

  if (tag.includes('unsupported')) {
    return ElectronReleaseChannel.unsupported;
  }

  // Must be a stable version, right?
  return ElectronReleaseChannel.stable;
}

export const enum VersionKeys {
  local = 'local-electron-versions',
  known = 'known-electron-versions'
}

/**
 * Retrieve Electron versions from localStorage.
 *
 * @param {VersionKeys} key
 * @param {() => Array<NpmVersion>} fallbackMethod
 * @returns {Array<NpmVersion>}
 */
function getVersions(
  key: VersionKeys, fallbackMethod: () => Array<NpmVersion>
): Array<NpmVersion> {
  const fromLs = window.localStorage.getItem(key);

  if (fromLs) {
    try {
      let result: Array<NpmVersion> = JSON.parse(fromLs);

      if (!isExpectedFormat(result)) {
        // Known versions can just be downloaded again.
        if (key === VersionKeys.known) {
          throw new Error(`Electron versions in LS does not match expected format`);
        }

        // Local versions are a bit more tricky and might be in an old format (pre 0.5)
        result = migrateVersions(result);
        saveLocalVersions(result);
      }

      return result;
    } catch (error) {
      console.warn(`Parsing local Electron versions failed, returning fallback method.`);
    }
  }

  return fallbackMethod();
}

/**
 * Save an array of GitHubVersions to localStorage.
 *
 * @param {VersionKeys} key
 * @param {Array<NpmVersion} versions
 */
function saveVersions(key: VersionKeys, versions: Array<NpmVersion>) {
  const stringified = JSON.stringify(versions);
  window.localStorage.setItem(key, stringified);
}

/**
 * Return both known as well as local versions.
 *
 * @returns {Array<NpmVersion>}
 */
export function getElectronVersions(): Array<ElectronVersion> {
  const known: Array<ElectronVersion> = getKnownVersions().map((version) => {
    return {
      ...version,
      source: ElectronVersionSource.remote,
      state: ElectronVersionState.unknown
    };
  });

  const local: Array<ElectronVersion> = getLocalVersions().map((version) => {
    return {
      ...version,
      source: ElectronVersionSource.local,
      state: ElectronVersionState.ready
    };
  });

  return [ ...known, ...local ];
}

/**
 * Add a version to the local versions
 *
 * @param {NpmVersion} input
 * @returns {Array<NpmVersion>}
 */
export function addLocalVersion(input: NpmVersion): Array<NpmVersion> {
  const versions = getLocalVersions();

  if (!versions.find((v) => v.localPath === input.localPath)) {
    versions.push(input);
  }

  saveLocalVersions(versions);

  return versions;
}

/**
 * Retrieves local Electron versions, configured by the user.
 *
 * @returns {Array<NpmVersion>}
 */
export function getLocalVersions(): Array<NpmVersion> {
  const versions = getVersions(VersionKeys.local, () => []);

  return versions;
}

/**
 * Saves local versions to localStorage.
 *
 * @param {Array<NpmVersion>} versions
 */
export function saveLocalVersions(versions: Array<NpmVersion | ElectronVersion>) {
  const filteredVersions = versions.filter((v) => {
    if (isElectronVersion(v)) {
      return v.source === ElectronVersionSource.local;
    }

    return true;
  });

  return saveVersions(VersionKeys.local, filteredVersions);
}

/**
 * Retrieves our best guess regarding the latest Electron versions. Tries to
 * fetch them from localStorage, then from a static releases.json file.
 *
 * @returns {Array<NpmVersion>}
 */
export function getKnownVersions(): Array<NpmVersion> {
  return getVersions(VersionKeys.known, () => require('../../static/releases.json'));
}

/**
 * Saves known versions to localStorage.
 *
 * @param {Array<NpmVersion>} versions
 */
export function saveKnownVersions(versions: Array<NpmVersion>) {
  return saveVersions(VersionKeys.known, versions);
}

/**
 * Tries to refresh our known versions and returns whatever we have
 * saved after.
 *
 * @export
 * @returns {Promise<Array<ElectronVersion>>}
 */
export async function getUpdatedElectronVersions(
): Promise<Array<ElectronVersion>> {
  try {
    await fetchVersions();
  } catch (error) {
    console.warn(`Versions: Failed to fetch versions`, { error });
  }

  return getElectronVersions();
}

/**
 * Fetch the latest known versions directly from npm.
 *
 * @returns {Promise<Array<NpmVersion>>}
 */
export async function fetchVersions() {
  const channels = [
    `https://registry.npmjs.org/electron`,        // stable, beta
    `https://registry.npmjs.org/electron-nightly` // nightly
  ];

  const output: Array<NpmVersion> = [];
  for (const channelUrl of channels) {
    const response = await window.fetch(channelUrl);
    const data = await response.json();
    const versions: Record<string, any> = data.versions;
    output.push(...Object.keys(versions).map((version) => ({ version })));
  }

  if (output && output.length > 0 && isExpectedFormat(output)) {
    console.log(`Fetched new Electron versions (Count: ${output.length})`);
    saveKnownVersions(output);
  }

  return output;
}

/**
 * Is the given array an array of versions?
 *
 * @param {Array<any>} input
 * @returns {boolean}
 */
export function isExpectedFormat(input: Array<any>): boolean {
  return input.every((entry) => !!entry.version);
}

/**
 * Migrates old versions, if necessary
 *
 * @param {Array<any>} input
 * @returns {Array<NpmVersion>}
 */
export function migrateVersions(input: Array<any> = []): Array<NpmVersion> {
  return input
    .filter((item) => !!item)
    .map((item) => {
      const { tag_name, name, url } = item;

      if (!tag_name || !name || !url) return null;

      return {
        version: tag_name,
        name,
        localPath: url
      };
    })
    .filter((item) => !!item) as Array<NpmVersion>;
}

export function isElectronVersion(
  input: NpmVersion | ElectronVersion
): input is ElectronVersion {
  return (input as ElectronVersion).source !== undefined;
}
