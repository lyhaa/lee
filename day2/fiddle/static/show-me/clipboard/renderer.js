const { clipboard } = require('electron')

const copyButton = document.querySelector('#copy')
const pasteButton = document.querySelector('#paste')
const textarea = document.querySelector('textarea')

copyButton.onclick = () => {
  clipboard.writeText('Hello from Electron!')
}

pasteButton.onclick = () => {
  textarea.value = clipboard.readText()
}
