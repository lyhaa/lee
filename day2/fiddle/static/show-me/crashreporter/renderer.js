const { crashReporter } = require('electron')

crashReporter.start({
  productName: 'YourName',
  companyName: 'YourCompany',
  submitURL: 'https://your-domain.com/url-to-submit',
  uploadToServer: true
})

setTimeout(() => {
  process.crash()
}, 2000)
