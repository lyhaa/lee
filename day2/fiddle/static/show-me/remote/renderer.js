const { remote } = require('electron')

// The BrowserWindow module can only be used from
// the main process, but thanks to the remote, we
// can create one from this renderer process.

const myWindow = new remote.BrowserWindow()
console.log(myWindow)

// Let's open developer tools
remote
  .getCurrentWebContents()
  .toggleDevTools()

console.log(remote)
