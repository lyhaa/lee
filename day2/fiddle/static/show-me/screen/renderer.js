const { screen } = require('electron')

const screens = screen.getAllDisplays()
const pre = document.querySelector('pre')

pre.innerText = JSON.stringify(screens, undefined, 2)
