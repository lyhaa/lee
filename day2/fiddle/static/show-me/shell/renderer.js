const { shell } = require('electron')

document.querySelector('#open-github').onclick = () => {
  shell.openExternal('https://github.com')
}

document.querySelector('#open-folder').onclick = () => {
  shell.showItemInFolder(__dirname)
}

document.querySelector('#beep').onclick = () => {
  shell.beep()
}
