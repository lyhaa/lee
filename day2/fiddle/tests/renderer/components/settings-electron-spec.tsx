import { mount, shallow } from 'enzyme';
import * as React from 'react';

import { ElectronVersionSource, ElectronVersionState } from '../../../src/interfaces';
import { ElectronSettings } from '../../../src/renderer/components/settings-electron';
import { ElectronReleaseChannel } from '../../../src/renderer/versions';
import { mockVersions } from '../../mocks/electron-versions';

describe('ElectronSettings component', () => {
  let store: any;

  beforeEach(() => {
    store = {
      version: '2.0.1',
      versions: { ...mockVersions },
      versionsToShow: [ ElectronReleaseChannel.stable, ElectronReleaseChannel.beta ],
      statesToShow: [ ElectronVersionState.ready, ElectronVersionState.downloading ],
      downloadVersion: jest.fn(),
      removeVersion: jest.fn(),
      updateElectronVersions: jest.fn(),
      toggleAddVersionDialog: jest.fn()
    };

    // Render all the states
    store.versions['2.0.2'].state = 'downloading';
    store.versions['2.0.1'].state = 'ready';
    store.versions['1.8.7'].state = 'unknown';
  });

  it('renders', () => {
    store.versions['3.0.0-nightly.1'] = {
      state: ElectronVersionState.ready,
      version: '3.0.0-nightly.1',
      source: ElectronVersionSource.local
    };

    store.versions['3.0.0'] = {
      state: ElectronVersionState.ready,
      version: '3.0.0',
      source: ElectronVersionSource.local
    };

    const wrapper = shallow(<ElectronSettings appState={store} />);

    expect(wrapper).toMatchSnapshot();
  });

  it('handles removing a version', async () => {
    store.versions['3.0.0-nightly.1'] = {
      state: ElectronVersionState.ready,
      version: '3.0.0-nightly.1',
      source: ElectronVersionSource.local
    };

    store.versions['3.0.0'] = {
      state: ElectronVersionState.ready,
      version: '3.0.0',
      source: ElectronVersionSource.local
    };

    const wrapper = mount(<ElectronSettings appState={store} />);

    wrapper
      .find('.electron-versions-table .bp3-button')
      .first()
      .simulate('click');

    expect(store.removeVersion).toHaveBeenCalledTimes(1);
  });

  it('handles downloading a version', async () => {
    store.versions = {
      '3.0.0': {
        state: ElectronVersionState.unknown,
        version: '3.0.0',
        source: ElectronVersionSource.remote
      }
    };

    store.statesToShow.push(ElectronVersionState.unknown);

    const wrapper = mount(<ElectronSettings appState={store} />);

    wrapper
      .find('.electron-versions-table .bp3-button')
      .first()
      .simulate('click');

    expect(store.downloadVersion).toHaveBeenCalledTimes(1);
  });

  it('handles the deleteAll()', async () => {
    const wrapper = shallow(<ElectronSettings appState={store} />);
    const instance = wrapper.instance() as any;
    await instance.handleDeleteAll();

    expect(store.removeVersion).toHaveBeenCalledTimes(2);
  });

  it('handles the downloadAll()', async () => {
    const wrapper = shallow(
      <ElectronSettings appState={store} />
    );
    const instance = wrapper.instance() as any;
    await instance.handleDownloadAll();

    expect(store.downloadVersion).toHaveBeenCalled();
  });

  describe('handleDownloadClick()', () => {
    it('kicks off an update of Electron versions', async () => {
      const wrapper = shallow(
        <ElectronSettings appState={store} />
      );
      const instance = wrapper.instance() as any;
      await instance.handleDownloadClick();

      expect(store.updateElectronVersions).toHaveBeenCalled();
    });
  });

  describe('handleAddVersion()', () => {
    it('toggles the add version dialog', () => {
      const wrapper = shallow(
        <ElectronSettings appState={store} />
      );
      const instance = wrapper.instance() as any;
      instance.handleAddVersion();

      expect(store.toggleAddVersionDialog).toHaveBeenCalled();
    });
  });

  describe('handleVersionChange()', () => {
    it('handles a new selection', async () => {
      const wrapper = shallow(
        <ElectronSettings appState={store} />
      );
      const instance = wrapper.instance() as any;
      await instance.handleStateChange({
        currentTarget: {
          id: ElectronVersionState.ready,
          checked: false
        }
      });

      await instance.handleStateChange({
        currentTarget: {
          id: ElectronVersionState.unknown,
          checked: true
        }
      });

      expect(store.statesToShow).toEqual([
        ElectronVersionState.downloading,
        ElectronVersionState.unknown
      ]);
    });
  });

  describe('handleChannelChange()', () => {
    it('handles a new selection', async () => {
      const wrapper = shallow(
        <ElectronSettings appState={store} />
      );
      const instance = wrapper.instance() as any;
      await instance.handleChannelChange({
        currentTarget: {
          id: ElectronReleaseChannel.stable,
          checked: false
        }
      });

      await instance.handleChannelChange({
        currentTarget: {
          id: ElectronReleaseChannel.nightly,
          checked: true
        }
      });

      expect(store.versionsToShow).toEqual([
        ElectronReleaseChannel.beta,
        ElectronReleaseChannel.nightly
      ]);
    });
  });
});
