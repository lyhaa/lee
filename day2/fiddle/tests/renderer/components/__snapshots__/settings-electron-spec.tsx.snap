// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ElectronSettings component renders 1`] = `
<div
  className="settings-electron"
>
  <h2>
    Electron Settings
  </h2>
  <Blueprint3.Callout>
    <Blueprint3.FormGroup
      label="Include Electron versions from these release channels:"
    >
      <Blueprint3.Checkbox
        checked={true}
        id="Stable"
        inline={true}
        label="Stable"
        onChange={[Function]}
      />
      <Blueprint3.Checkbox
        checked={true}
        id="Beta"
        inline={true}
        label="Beta"
        onChange={[Function]}
      />
      <Blueprint3.Checkbox
        checked={false}
        id="Nightly"
        inline={true}
        label="Nightly"
        onChange={[Function]}
      />
      <Blueprint3.Checkbox
        checked={false}
        id="Unsupported"
        inline={true}
        label="Unsupported"
        onChange={[Function]}
      />
    </Blueprint3.FormGroup>
    <Blueprint3.FormGroup
      label="Include Electron versions that are:"
    >
      <Blueprint3.Checkbox
        checked={true}
        id="downloading"
        inline={true}
        label="Downloading"
        onChange={[Function]}
      />
      <Blueprint3.Checkbox
        checked={true}
        id="ready"
        inline={true}
        label="Downloaded"
        onChange={[Function]}
      />
      <Blueprint3.Checkbox
        checked={false}
        id="unknown"
        inline={true}
        label="Not Downloaded"
        onChange={[Function]}
      />
    </Blueprint3.FormGroup>
  </Blueprint3.Callout>
  <br />
  <Blueprint3.Callout>
    <Blueprint3.ButtonGroup
      fill={true}
    >
      <Blueprint3.Button
        icon="numbered-list"
        onClick={[Function]}
        text="Update Electron Release List"
      />
      <Blueprint3.Button
        disabled={false}
        icon="download"
        onClick={[Function]}
        text="Download All Versions"
      />
      <Blueprint3.Button
        disabled={false}
        icon="trash"
        onClick={[Function]}
        text="Delete All Downloads"
      />
      <Blueprint3.Button
        icon="document-open"
        onClick={[Function]}
        text="Add Local Electron Build"
      />
    </Blueprint3.ButtonGroup>
    <HTMLTable
      className="electron-versions-table"
      striped={true}
    >
      <thead>
        <tr>
          <th>
            Version
          </th>
          <th>
            Status
          </th>
          <th
            className="action"
          >
            Action
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          key="3.0.0"
        >
          <td>
            3.0.0
          </td>
          <td>
            <span>
              <Blueprint3.Icon
                icon="box"
              />
               
              Downloaded
            </span>
          </td>
          <td
            className="action"
          >
            <Blueprint3.Button
              fill={true}
              icon="trash"
              onClick={[Function]}
              small={true}
              text="Remove"
            />
          </td>
        </tr>
        <tr
          key="2.0.2"
        >
          <td>
            2.0.2
          </td>
          <td>
            <span>
              <Blueprint3.Icon
                icon="cloud-download"
              />
               
              Downloading
            </span>
          </td>
          <td
            className="action"
          >
            <Blueprint3.Button
              disabled={true}
              fill={true}
              icon="cloud-download"
              loading={true}
              small={true}
              text="Downloading"
            />
          </td>
        </tr>
        <tr
          key="2.0.1"
        >
          <td>
            2.0.1
          </td>
          <td>
            <span>
              <Blueprint3.Icon
                icon="box"
              />
               
              Downloaded
            </span>
          </td>
          <td
            className="action"
          >
            <Blueprint3.Button
              fill={true}
              icon="trash"
              onClick={[Function]}
              small={true}
              text="Delete"
            />
          </td>
        </tr>
      </tbody>
    </HTMLTable>
  </Blueprint3.Callout>
</div>
`;
